### ISSUE TITLE
 Add new 'chef_client_config' resource

### ISSUE THREAD

**Issue by mattray** [chef/chef#10094]
Configuring the Chef Infra Client requires having a config file (usually `/etc/chef/client.rb`) that is used to manage the agent. This resource will set up the config file from a template and reload the configuration for the current chef-client run.

# Motivation
As a user, I want a built-in resource for configuring the Chef Infra Client configuration without depending on an external cookbook. Most users use the [chef-client](https://github.com/chef-cookbooks/chef-client) `config` recipe for this, but when combined with the new `chef_client_cron`, `chef_client_scheduled_task`, or `chef_client_systemd_timer` they should be able to stop using the `chef-client` cookbook completely.

# Specification
By default this resource will write out the `/etc/chef/client.rb` configuration file from a template and reload the configuration for the current chef-client run.

This resource will expose the [documented settings](https://docs.chef.io/config_rb_client/#settings) for the `client.rb` as properties with defaults provided where applicable. Directories will be created for logs and the configuration file. Unlike the [chef-client::config recipe](https://github.com/chef-cookbooks/chef-client/blob/master/recipes/config.rb), hashes of attributes will not be supported. Settings that may be strings or symbols (`log_level`, `ssl_verify_mode`, etc.) may be entered as their Ruby type.

## Example
```
chef_client_config 'Chef Infra Client configuration file' do
    path '/etc/chef/client.rb'
    chef_license "accept"
    log_location "/var/log/chef/client.log"
    log_level :debug
    ohai_disabled_plugins ["dmi"]
    mode '0644'
    action :create
end
```

## Actions
- `:create`: Default. Creates the config file. If the file already exists (but does not match), update that file to match.
- `:create_if_missing`: Create the file only if the file does not exist. When the file exists, nothing happens.
- `:delete`: Delete the config file.

## Properties
The [documented settings](https://docs.chef.io/config_rb_client/#settings) for the `client.rb` will be available as properties with defaults provided where applicable. In addition to these settings, the relevant [file](https://docs.chef.io/resources/file/) properties will be available. 

###  chef_client_config Properties
- `path`: String, defaults to the name if not specified.
- `content`- String, appended to end of the template.
- `reload`: true, false, whether to reload the configuration for the current chef-client run. Defaults to `true`. 

### client.rb-derived Properties
- `add_formatter`
- `authentication_protocol_version`
- `automatic_attribute_blacklist`
- `automatic_attribute_whitelist`
- `cache_path`
- `checksum_path`
- `chef_guid`
- `chef_license`
- `chef_repo_path`
- `chef_server_url`
- `chef_zero.enabled`
- `chef_zero.port`
- `clear_gem_sources`
- `client_fork`
- `client_key`
- `client_registration_retries`
- `client_d_dir`
- `cookbook_path`
- `cookbook_sync_threads`
- `data_bag_decrypt_minimum_version`
- `data_bag_path`
- `data_collector.server_url`
- `data_collector.token`
- `data_collector.mode`
- `data_collector.raise_on_failure`
- `default_attribute_blacklist`
- `default_attribute_whitelist`
- `diff_disabled`
- `diff_filesize_threshold`
- `diff_output_threshold`
- `disable_event_logger`
- `enable_reporting`
- `enable_reporting_url_fatals`
- `enable_selinux_file_permission_fixup`
- `encrypted_data_bag_secret`
- `enforce_path_sanity`
- `environment`
- `environment_path`
- `exit_status`
- `file_atomic_update`
- `file_cache_path`
- `file_staging_uses_destdir`
- `fips`
- `force_formatter`
- `force_logger`
- `ftp_proxy`
- `ftp_proxy_pass`
- `ftp_proxy_user`
- `chef_group`- this is the `group` in the configuration file that owns a process, but it is changed so it does not conflict with the file `group` property.
- `gem_installer_bundler_options`
- `http_proxy`
- `http_proxy_pass`
- `http_proxy_user`
- `http_retry_count`
- `http_retry_delay`
- `https_proxy`
- `https_proxy_pass`
- `https_proxy_user`
- `interval`
- `json_attribs`
- `listen`
- `local_key_generation`
- `local_mode`
- `lockfile`
- `log_level`
- `log_location`
- `minimal_ohai`
- `named_run_list`
- `no_lazy_load`
- `no_proxy`
- `node_name`
- `node_path`
- `normal_attribute_blacklist`
- `override_attribute_blacklist`
- `normal_attribute_whitelist`
- `override_attribute_whitelist`
- `pid_file`
- `policy_group`
- `policy_group_path`
- `policy_name`
- `policy_path`
- `recipe_url`
- `rest_timeout`
- `role_path`
- `rubygems_url`
- `run_lock_timeout`
- `script_path`
- `skip_gem_metadata_installation`
- `splay`
- `stream_execute_output`
- `show_download_progress`
- `download_progress_interval`
- `ssl_ca_file`
- `ssl_ca_path`
- `ssl_client_cert`
- `ssl_client_key`
- `ssl_verify_mode`
- `trusted_certs_dir`
- `umask`
- `use_policyfile`
- `user`
- `validation_client_name`
- `validation_key`
- `verbose_logging`
- `verify_api_cert`
- `whitelist`

### file-derived Properties
- `atomic_update`
- `backup`
- `checksum`
- `force_unlink`
- `group`
- `inherits`
- `manage_symlink_source`
- `mode`
- `owner`
- `rights`
- `verify`
### Common Resource Properties
- `compile_time`
- `ignore_failure`
- `retries`
- `retry_delay`
- `sensitive`

**Comment by lamont-granquist** [comment]
That is a huge pile of properties, and the concerns that I have about this is that it means a huge amount of ongoing bugfixes and complains and maintenance since that will be a never ending API target.

As new configs are added or removed or extended so that the validation here breaks, then it'll just generate new bugs and complaints to a team that isn't staffed adequately to handle it.

Sort of in a related way it seems not very useful to export all of the file properties, since atomic_updates and other properties should not need to be tweaked.  Most of those seem to be exposing APIs that nobody should ever need to use to control the config setting.  It should definitely not expose them due to inheritance from the file resource and should be written with composition.  Some of them like mode/owner/rights makes more sense.

Oh there's also obvious namespaceing issues here where `Chef::Config[:sensitive]` or `Chef::Config[:compile_time]` could never be set via a correctly named property, which additionally makes the idea of single-property-per-every-setting not very attractive.  This approach more or less recreates the problems with method_missing.

**Comment by lamont-granquist** [comment]
Its also possible that this should go as a set of common properties that behave the same across all the chef_client resources.  Between config/path/content/reload and the mode/groups/owner/rights properties that you do need, I can buy the argument that you've got enough there for a different resource, but at the same time if you keep it all together it keeps it more discoverable for users and they only have to use one thing instead of two.  Would clearly want to use a resource partial in that case though.  I'd like to at least see the defense of why one was picked over the other though because I'm split very 50/50 in my head about which is best for end users.

**Comment by tas50** [comment]
I think we'd want to come up with a subset that would satisfy 90% of users 90% of the time and then use those. It's probably 15 or so.

**Comment by lamont-granquist** [comment]
Its gonna generate bug reports to fill out the entire rest of the API, we'll see those dribble in one by one.

**Comment by lamont-granquist** [comment]
Also matches what we do with stuff like package resources with the `options` flag and just let people who have one off use cases inject arbitrary things into the command.  Then we build out the API as we hit use cases where we actually need to handle specific options out of that property (because we need to mangle the command that gets the list of available/installed packages differently and specially for options that eventually wind up being passed into the install command or whatever -- thats a good use case for peeling off a new property).

We do have resources that spam the whole config file into the property list, but those aren't well maintained at all, I can probably find years-old bug reports about then, and if they do start being used more going forwards, we're probably going to wind up engineering a "toss-everything-else-in-here" kind of property to them.

**Comment by tas50** [comment]
Yeah ideally we wouldn't make it look like yum_repository which has every option in there. I think a options property would do nicely for folks with use cases outside the norm. Most people are probably setting the same 12 options though.

**Comment by mattray** [comment]
I'm more than happy to start with a pared-down subset covering 90% of the use cases and having an `options` hash. If that's the path to eventual inclusion I'll start up an MVP cookbook for feedback/POC.

**Comment by mattray** [comment]
This was released with Chef Client 16.6: https://docs.chef.io/resources/chef_client_config/
